### ISSUE TITLE
slate-history `shouldMerge` is not accurate

### ISSUE THREAD

**Issue by githoniel** [ianstormtaylor/slate#3838]
#### Do you want to request a _feature_ or report a _bug_?

BUG

#### What's the current behavior?

![xx](https://user-images.githubusercontent.com/17848159/91134312-4b620800-e6e0-11ea-906b-7906d473f9cf.gif)

1. input `dragTest1`, press enter
2. input `dragTest2`, press enter
3. make i reverse selection of `dragTest2`
4. drag it after `dragTest1`
5. press ctrl+ z to undo
6. we will get `dragTest1` but expect `dragTest1\ndragTest2`

Slate: 0.58.4
Browser: Chrome / Safari / Firefox / Edge
OS: Mac / Windows / Linux / iOS / Android

#### How does this happen

after debugging, I found this is caused by wrong history merge.

after 3, the current history is 

![image](https://user-images.githubusercontent.com/17848159/91140833-5453d900-e6e2-11ea-9b2a-e04a0777abc0.png)

then we drag as 4, it will fire two `onDOMBeforeInput` event as expect, input type will be `deleteByDrag` and `insertFromDrop`. 

this bug is caused by `deleteByDrag`

```ts
if (!type.startsWith('delete') || type.startsWith('deleteBy')) {
  const [targetRange] = event.getTargetRanges()

  if (targetRange) {
    const range = ReactEditor.toSlateRange(editor, targetRange)

    if (!selection || !Range.equals(selection, range)) {
      Transforms.select(editor, range)
    }
  }
}
```

`targetRange` will alway be ordinal regardless of current selection's direction. so we will apply an `set_selection` operation, which will cause all opertion after will by merge into previous history.

![image](https://user-images.githubusercontent.com/17848159/91140172-22427700-e6e2-11ea-8fee-c8c74186bf12.png)

#### how to fix

I think change `Range.equals` to `isRangeEqualIgnoreDirection` is not the right way, we should controll whether to merge history or create a new history manually in `slate-react`.

`slate-history` provide a api `withouMerging` seem weird

```ts
  /**
   * Apply a series of changes inside a synchronous `fn`, without merging any of
   * the new operations into previous save point in the history.
   */

  withoutMerging(editor: HistoryEditor, fn: () => void): void {
    const prev = HistoryEditor.isMerging(editor)
    MERGING.set(editor, false)
    fn()
    MERGING.set(editor, prev)
  },
```

does restore previous merge state make sense after we have alreay insert a new history array?

I think we should refact it to `setMerging(willMerge: boolean): void` and `slate-react` need to use this api to make accurate history merge

**Comment by YaoKaiLun** [comment]
Maybe you can do like this
```
HistoryEditor.withoutSaving(editor, () => {
   Transforms.select(editor, range)
});
```

**Comment by verngutz** [comment]
Here is a simpler case where this bug occurs:
1. Type any text into an empty editor.
2. Select any part of the text in the backward direction.
3. Type any text (replace the selection).
4. Hit undo. The text typed in Step 1 also gets undone.

The problem is not in the `shouldMerge` function. In the current implementation, a HistoryEditor will always merge the new operation if the BaseEditor has any pending unflushed operations. This becomes a problem when any `set_selection` operation is applied (e.g. called by `onDOMBeforeInput`) before any saved operations are applied. The operations are flushed only after all operations in the batch are applied. While applying the saved operation, `set_selection` operations exist in the `operations` array of the BaseEditor and therefore, HistoryEditor assumes the next non-`set_selection` operation should merge with it. But `set_selection` operations are no longer saved after #4717, so the next operation gets wrongly merged with the previous batch.

I think the fix should be to filter out all `set_selection` operations from the `operations` array before doing anything else in HistoryEditor's `apply` function. I'm just not sure what else this affects or breaks.
